src/main/java/seedu/duke/cli/Parser.java

```diff
@@ -0,0 +1,183 @@
+package seedu.duke.cli;
+
+import seedu.duke.cli.annotations.Argument;
+import seedu.duke.cli.annotations.CommandConstructor;
+import seedu.duke.cli.commands.ByeCommand;
+import seedu.duke.cli.commands.DeadlineCommand;
+import seedu.duke.cli.commands.DeleteCommand;
+import seedu.duke.cli.commands.DoneCommand;
+import seedu.duke.cli.commands.EventCommand;
+import seedu.duke.cli.commands.FindCommand;
+import seedu.duke.cli.commands.ListCommand;
+import seedu.duke.cli.commands.TodoCommand;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.HashMap;
+import java.util.regex.Pattern;
+
+public class Parser {
+    private final static DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ofPattern("d/M/uuuu HH:mm");
+    private static HashMap<String, ConstructorCache> constructors;
+
+    static {
+        constructors = new HashMap<>();
+        addConstructors(ListCommand.class);
+        addConstructors(ByeCommand.class);
+        addConstructors(DoneCommand.class);
+        addConstructors(TodoCommand.class);
+        addConstructors(EventCommand.class);
+        addConstructors(DeadlineCommand.class);
+        addConstructors(DeleteCommand.class);
+        addConstructors(FindCommand.class);
+    }
+
+    // This class is to be used statically only.
+    private Parser() {
+    }
+
+    @SuppressWarnings("unchecked")
```

Comment:
Could I ask why it was done this way? Kind of confusing for my untrained eye haha

Reply:
Hmm, I didn't want to have a giant switch/case or repeat the argument parsing code.. This was the first idea that came to mind I guess. Actually attributes aren't strictly needed here, you could use Builder subclasses or something like that to achieve the same result, but that would result in more boilerplate, I think.